collector/pg_replication_slots.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2022 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	registerCollector("replication_slot", defaultEnabled, NewPGReplicationSlotCollector)
+}
+
+type PGReplicationSlotCollector struct {
+	log log.Logger
+}
+
+func NewPGReplicationSlotCollector(logger log.Logger) (Collector, error) {
+	return &PGReplicationSlotCollector{log: logger}, nil
+}
+
+var pgReplicationSlot = map[string]*prometheus.Desc{
```

Comment:
The thing about putting them in the struct of each collector is that it makes the New func a bit bloated in my opinion. Putting them in the struct would keep the package namespace more clean though. If they're in the struct, the probe endpoint will recreate them on every request the way that it is coded today, so that might increase the memory footprint of the collector. I think if we change it to struct scope that we should do that at a later time after we refactor the probe endpoint to not re-init the collector package on every request.

Reply:
>  If they're in the struct, the probe endpoint will recreate them on every request the way that it is coded today, so that might increase the memory footprint of the collector.

Ahh, right, that's a good reason to keep them as package vars outside of the struct. Reduced GC churn.